plugins/woocommerce/tests/legacy/unit-tests/email/emails.php

```diff
@@ -101,4 +102,203 @@ public function test_headers_with_enabled_email_improvements() {
+				return $replace;
+			},
+			10
+		);
+		add_filter(
+			'woocommerce_email_format_string_find',
+			function ( $find ) {
+				$find['customer-name'] = '{customer_name}';
+				return $find;
+			},
+			10
+		);
+
+		$result = $email->format_string( 'Hello {customer_name}' );
+		$this->assertEquals( 'Hello Jane Smith', $result );
+
+		// Clean up filters.
+		remove_all_filters( 'woocommerce_email_format_string_replace' );
+		remove_all_filters( 'woocommerce_email_format_string_find' );
+	}
+
+	/**
+	 * Test legacy blogname placeholder with filter override in format_string.
+	 * Note: blogname is a special case in that it is not a placeholder, but a string when in the scope of the filters.
+	 * So for users wanting to support this replacement string, they need to use both filters.
+	 */
+	public function test_format_string_legacy_blogname_filter_override() {
+		$email = new WC_Email();
+
+		add_filter(
+			'woocommerce_email_format_string_replace',
+			function ( $replace ) {
+				$replace['blogname'] = 'My Custom Shop Name';
+				return $replace;
+			},
+			10
+		);
+		add_filter(
+			'woocommerce_email_format_string_find',
+			function ( $find ) {
+				$find['blogname'] = '{blogname}';
+				return $find;
+			},
+			10
+		);
+
+		$this->assertEquals(
+			'Welcome to My Custom Shop Name!',
+			$email->format_string( 'Welcome to {blogname}!' ),
+			'Legacy filters should be able to override the default blogname replacement but cannot due to a bug'
+		);
+
+		// Clean up filters.
+		remove_all_filters( 'woocommerce_email_format_string_replace' );
+		remove_all_filters( 'woocommerce_email_format_string_find' );
+	}
+
+	/**
+	 * Test main filter override in format_string.
+	 */
+	public function test_format_string_main_filter_override() {
+		$email = new WC_Email();
+
+		add_filter(
+			'woocommerce_email_format_string',
+			function () {
+				return 'Completely overridden';
+			},
+			10
+		);
+
+		$this->assertEquals(
+			'Completely overridden',
+			$email->format_string( 'Original text with {customer_name}' )
+		);
+
+		// Clean up.
+		remove_all_filters( 'woocommerce_email_format_string' );
+	}
+
+	/**
+	 * Test empty string handling in format_string.
+	 */
+	public function test_format_string_empty_string() {
+		$email = new WC_Email();
+		$this->assertEquals( '', $email->format_string( '' ) );
+	}
+
+	/**
+	 * Test string without placeholders in format_string.
+	 */
+	public function test_format_string_no_placeholders() {
+		$email = new WC_Email();
+		$this->assertEquals(
+			'Just a regular string',
+			$email->format_string( 'Just a regular string' )
+		);
+	}
+
+	/**
+	 * Test partial/invalid placeholders in format_string.
+	 */
+	public function test_format_string_partial_placeholders() {
+		$email               = new WC_Email();
+		$email->placeholders = array(
+			'{test}' => 'TEST',
+		);
+		$this->assertEquals(
+			'Partial {test TEST {test_invalid}',
+			$email->format_string( 'Partial {test TEST {test_invalid}' )
```

Comment:
```suggestion
			$email->format_string( 'Partial {test {test} {test_invalid}' )
```

Reply:
Not a typo, but the change makes sense! Thanks :)